src/int/div.rs

```diff
@@ -0,0 +1,314 @@
+//! [`Int`] division operations.
+
+use core::ops::Div;
+
+use subtle::{Choice, CtOption};
+
+use crate::{CheckedDiv, ConstChoice, Int, NonZero, Uint};
+
+impl<const LIMBS: usize> Int<LIMBS> {
+    #[inline]
+    /// Base checked_div_mod operation.
+    /// Given `(a, b)` computes the quotient and remainder of the absolute
+    /// Note: this operation rounds towards zero, truncating any fractional part of the exact result.
+    fn checked_div_mod(
+        &self,
+        rhs: &NonZero<Self>,
+    ) -> (Uint<{ LIMBS }>, Uint<{ LIMBS }>, ConstChoice) {
+        // Step 1: split operands into signs, magnitudes and whether they are zero.
+        let (lhs_sgn, lhs_mag) = self.sign_and_magnitude();
+        let (rhs_sgn, rhs_mag) = rhs.sign_and_magnitude();
+
+        // Step 2. Determine if the result should be negated.
+        // This should be done if and only if lhs and rhs have opposing signs.
+        // Note: if the lhs is zero, the resulting magnitude will also be zero. Negating zero,
+        // however, still yields zero, so having a truthy `negate_result` in that scenario is OK.
+        let opposing_signs = lhs_sgn.xor(rhs_sgn);
+
+        // Step 3. Construct result
+        // safe to unwrap since rhs is NonZero.
+        let (quotient, remainder) = lhs_mag.div_rem(&NonZero::new(rhs_mag).unwrap());
+
+        (quotient, remainder, opposing_signs)
+    }
+
+    /// Perform checked division, returning a [`CtOption`] which `is_some` only if the rhs != 0.
+    /// Note: this operation rounds towards zero, truncating any fractional part of the exact result.
+    pub fn checked_div(&self, rhs: &Self) -> CtOption<Self> {
+        NonZero::new(*rhs).and_then(|rhs| {
+            let (quotient, _, opposing_signs) = self.checked_div_mod(&rhs);
+            Self::new_from_sign_and_magnitude(opposing_signs, quotient).into()
+        })
+    }
+
+    /// Perform checked division, returning a [`CtOption`] which `is_some` only if the rhs != 0.
+    /// Note: this operation rounds down.
+    ///
+    /// Example:
+    /// ```
+    /// use crypto_bigint::I128;
+    /// assert_eq!(
+    ///     I128::from(8).checked_div_floor(&I128::from(3)).unwrap(),
+    ///     I128::from(2)
+    /// );
+    /// assert_eq!(
+    ///     I128::from(-8).checked_div_floor(&I128::from(3)).unwrap(),
+    ///     I128::from(-3)
+    /// );
+    /// assert_eq!(
+    ///     I128::from(8).checked_div_floor(&I128::from(-3)).unwrap(),
+    ///     I128::from(-3)
+    /// );
+    /// assert_eq!(
+    ///     I128::from(-8).checked_div_floor(&I128::from(-3)).unwrap(),
+    ///     I128::from(2)
+    /// )
+    /// ```
+    pub fn checked_div_floor(&self, rhs: &Self) -> CtOption<Self> {
```

Comment:
do we need this method?
I think one method that rounds towards zero it sufficient (so it works like regular primitive int division).

Reply:
Do _we_ need it? No, probably not. Do _I_ need it? Yes, yes I do :)

I need this operation as well as `div_mod_floor` for a project I'm working on (that op is one of the reasons I'm implementing this right now :see_no_evil:). IIUC, the definition I am using here aligns with `num_bigint::BigInt::div_mod_floor`, which I'm trying to replace to make my algorithm constant time.

This function, of course, differs slightly from  `Int::checked_div` in that it rounds down, instead of to zero.

Reply:
@erik-3milabs out of curiosity, if this is not a secret, what do you need the `floor_div` for?

Reply:
I need `Int` support to implement class-group cryptography. In one or two of the sub-algorithms, I specifically need floor-div.

Reply:
This is the reason I am asking because you mentioned CL before.
I have implemented Class Groups crypto (BICYCL paper and original Castagnos Laguillaumie paper) and didn't need this method so this got me thinking maybe I am doing something wrong :/.
BTW. Elements of group in my impl are represented in both binary quadratic forms and ideals (for compactness).

Reply:
Also, I'm curious: what did you do with the `xgcd` s in the algorithms? This repository does not provide an implementation for that...

Reply:
@erik-3milabs nice :).
As for gcdx I don't have a const time implementation yet, but I was hoping to use Berstein Yung for that. If it can be done for polynomials I so no reason why it wouldn't work for integers. But fair enough gcdx is still not constant time in my implementation :(.

Reply:
Ohh, I also believe that binary extended gcd can be adopted to be const time.

Reply:
Take a look at iacr 2020 972.
I plan to try do that constant time I Berstein Yang redness unusable for some reasons.

Reply:
Binary GCD may be of interest in `crypto-primes` as well (although possibly not a constant time implementation) /cc @fjarri @dvdplm

Reply:
We have a somewhat specific case where one argument is always small, so the general one provided by `crypto-bigint` is an overkill. @dvdplm added a binary GCD manually recently in https://github.com/entropyxyz/crypto-primes/pull/54

Reply:
@lleoha I was eyeing the 2020/972 paper as well :smirk: Still experimenting with it.